src/db.c

```diff
@@ -1704,9 +1704,13 @@ int getKeysUsingKeySpecs(struct redisCommand *cmd, robj **argv, int argc, int se
         keySpec *spec = cmd->key_specs + j;
         serverAssert(spec->begin_search_type != KSPEC_BS_INVALID);
         /* Skip specs that represent channels instead of keys */
-        if (spec->flags & (CMD_KEY_CHANNEL) && !(search_flags & GET_KEYSPEC_INCLUDE_CHANNELS)) {
+        if ((spec->flags & CMD_KEY_CHANNEL) && !(search_flags & GET_KEYSPEC_INCLUDE_CHANNELS)) {
             continue;
         }
+        /* Handle incomplete specs */
+        if (spec->find_keys_type & CMD_KEY_INCOMPLETE) {
+            goto invalid_spec;
+        }
```

Comment:
i think this `goto` should be at the bottom of the loop (just before the label).
so that if we have a valid key-spec (with `range` or `keyword`), and the user indicated he wants us to extract what's possible. we don't skip the spec.

alternatively, we can add an if `!GET_KEYSPEC_RETURN_PARTIAL` check on this goto.